### PR TITLE
Map with multiple inputs

### PR DESCRIPTION
**Pull request by dvd101x**
Hi, this is intended to add multiple inputs to the function map.

Currently this includes:
- Mapping multiple arrays or matrices
- Broadcasts the input matrices if needed
- Implements the callback function with multiple inputs
- Optionally the callback function can have the broadcasted index and all the broadcasted arrays or matrices.
- Tests for the new functionality
- Fixes an issue with the current map transform function missing the datatype of the matrix.

It's missing:
- The transfer function for the multiple map functionality
- Tests for the transform function new functionality
- Documentation

Could you please give me advice on a recommended way to implement the transform function?

I have tried or studied the following alternatives:
- Create the map function inside transfer, but had some issues with the callback function.
- Implementing the full logic of the main function, as some other transform functions, which I think could work but I don't know if it's ok. 
- Try to use the current transfer implementation and build on top on that with refer to self.

At this point I'm stuck so any pointers are greatly appreciated.

this is related to #3196

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks for working this idea out in a PR David!

About the transform function: I think the transform function should only be a small wrapper around the actual implementation, and only adjust the index in the callback function from zero based to one based. So I would expect that we have to adjust the current (recursive) map function:

https://github.com/josdejong/mathjs/blob/develop/src/expression/transform/map.transform.js#L63-L66

from `recurse(child, index.concat(i + 1))` into something like `recurse(child1, child2, index.concat(i + 1))`, but then probably something more generic like `recurse(childs, index.concat(i + 1))`. Did you gave that a try? If so, what issues did you came across when trying it?

**Comment by dvd101x** [comment]
Hi, thanks for the advice!

Currently `recurse` works for me, as it can recurse over the first array (broadcasted) and the callback function can take the argument of the rest (also broadcasted, thus the same size). So instead of changing `recurse`, it makes a special callback to make it work with the current `recurse`. It's not a pure function but it works. I have that logic on the main function, but I don't know if I should bring that code to the transform function. Your idea certainly looks cleaner.

Some issue that I found is that `map.transform` practically duplicates the code in `map` except that it adds one to the index, thus it's not a wrapper around the actual implementation. I could continue in that direction by copying the code but I'm not sure about it since the code is now longer.

By looking at other transform functions, I see in some cases those import the original function first, instead of duplicating the code. In the case of `map.tranform` why does it need to duplicate the code in `map`?

**Comment by dvd101x** [comment]
I think I'm on track again.

I'm working on a way to use the create map. My plan is to make a new callback that when the callback is a typed function it makes a new typed callback that for each signature changes the function associated with the signature to be zero based depending on the number of arguments of the function.

It might take a while but I think it might work. Just to let you know there is some progress.

**Comment by dvd101x** [comment]
Hi Jos,

I need your guidance again. What's a good way to get the number of arguments of `f` in the following example?
```js
const f = math.evaluate("f(x, index) = index[1]")
```

I tried with `f.signatures['any,any'].length` but gives me **0** when I expect **2**.

**Comment by josdejong** [comment]
Thanks for the updates. 

You indeed cannot safely use `f.signatures['any,any'].length`. But if you have `'any,any'`, you could do `'any,any'.split(',').length` for example? Not sure why you need it but I guess I'll see it in action in the PR :)

On a side note: since this is not easy to solve, let's stay critical: if this feature is very complex to implement (and consecutively: hard to maintain), let's think through if that is worth it.

**Comment by dvd101x** [comment]
Thanks Jos!,

I understand.

The main functionality is already working, it's currently just an issue with trying to fix the transform function without copying the logic and for that I think I need to transform the callback function with the right number of arguments. Maybe it makes more sense to continue copying the logic with a modification for the index than making new logic for manipulating the callback.

I don't think I'm far from having a working proposal, so we can decide then.

It has some similarities with julias dot notation for map and broadcast. Maybe some common ground can be found.
https://julialang.org/blog/2017/01/moredots/

**Comment by josdejong** [comment]
Sounds promising!

**Comment by dvd101x** [comment]
Hi Jos, it's ready for review.

**Comment by josdejong** [comment]
Thanks David! I'll look into it next week (I'm not going to make it this week, sorry).

**Comment by dvd101x** [comment]
Don't worry, thank you so much, I will chill for a bit.

It's missing some tests to pass the code coverage, update the embedded docs, maybe something related to types. But I will wait for your review.

**Comment by josdejong** [comment]
I've had a look at the multi map implementation, and it looks well taken care of, the logic is easy to follow. Nice job 👍 

Feedbacks:

1. I think there are some extra unit tests needed: 
    1. validate that the callback also passes the current index (array) and parent arrays
    2. in case of the `map.transform` ensure that the passed indices are one based instead of zero based
2. We need to update the documentation of the function and I think add a section or example to [the page about matrices](https://mathjs.org/docs/datatypes/matrices.html)
3. We need to update the type definitions
4. I expect that using multi map introduces a performance overhead, but it would be nice if we do the best we can. I tried the following test to see how it performs, and I have the feeling that it should be possible to improve on it: 

    ```js
    const a = math.ones(1000, 1000)
	
    console.time('add')
    math.add(a, a)
    console.timeEnd('add')
	// 18 milliseconds

    console.time('add via map')
    math.map(a, a, math.addScalar)
    console.timeEnd('add via map')
	// 1669 milliseconds
    ```
    Do you know where the bottleneck is and do you have ideas to improve on this? (could be a separate PR)

**Comment by dvd101x** [comment]
Thanks Jos!

I will look into fixing the issues.

In the meantime regarding performance. It might be related to map and the use of subset on each iteration.

By testing on the current version (13.0.3)

```js
const a = math.ones(1000, 1000)

console.time('add')
math.add(a, a)
console.timeEnd('add')
// 18 milliseconds

console.time('add 1 via map')
math.map(a, (av) => math.addScalar(av, 1))
console.timeEnd('add via map')
// 137 milliseconds

console.time('add via map with index')
math.map(a, (av, ai) => math.addScalar(av, a.subset(math.index(...ai))))
console.timeEnd('add via map with index')
// 1043 milliseconds
```

The logic followed on the third example is similar to the one implemented on the PR minus broadcasting and the effect on performance of subset can be shown.

I think there would be no need to make a subset, but some fundamental changes must be done on the function `_map` to be able to skip the last dimension and provide all elements of the last dimension to the callback.

```js
const A = [1, 1]
const B = [2, 2]

joinedArrays = [[1, 2], [1, 2]] // using some kind of f(A, B)

joinedArrays.map(values => {
    return math.addScalar(...values)
}) // [3, 3]
```

This includes a trade off between speed and memory allocation. There might be ways to do both broadcasting and map (multiple) without the need of filling matrices with repeated data, maybe we can come up with something.

**Comment by dvd101x** [comment]
Just found that the performance improves by using `.get` instead of `.subset`. Will look into it.

```js
const a = math.ones(1000, 1000)

console.time('add')
math.add(a, a)
console.timeEnd('add')
// 18 milliseconds

console.time('add 1 via map')
math.map(a, (av) => math.addScalar(av, 1))
console.timeEnd('add 1 via map')
// 137 milliseconds

console.time('add via map with subset')
math.map(a, (av, ai) => math.addScalar(av, a.subset(math.index(...ai))))
console.timeEnd('add via map with subset')
// 1043 milliseconds

console.time('add via map with get')
math.map(a, (av, ai) => math.addScalar(av, a.get(ai)))
console.timeEnd('add via map with get')
// 155 milliseconds
```

**Comment by josdejong** [comment]
Sounds promising!

**Comment by dvd101x** [comment]
Hi Jos,

I was testing performance and found another issue that I'm still resolving, but I want to bring a topic to your attention in the meantime.

By testing on the current v13.0.3 found something interesting.

| Method | Chrome [ms] | Edge [ms]  | Firefox [ms] |
| -------|--------------------|-------------------|---------------------|
| add    | 62 | 59 | 1373 |
| matrix map with subset |1312 | 1210 | 27699 |
| matrix map with get | 240 | 235 | 6191 |
| flat map | 108 | 110 | 1477 |

In these cases by avoiding recursion the performance improves. The method I tested was flattening the whole array, applying a map and then unflattening. Even with the overhead of all the conversions it seems like an improvement. Specially for Firefox.

This implies that a conversion for the index must be applied in the cases where the callback includes the index, but it doesn't seem like a big deal.

So I was thinking, what if the matrices were flat under the hood, to avoid both; recursion and flattening.

By doing some research I think numpy arrays work in that way under the hood.

https://numpy.org/devdocs/reference/arrays.ndarray.html#internal-memory-layout-of-an-ndarray


I'm not sure if this is a good idea, but I wanted to share it with you.

## reference code
```js
math.config({
  randomSeed: 100
})
const A = math.random([1000, 1000])
const B = math.random([1000, 1000])

console.time('add')
const arrayAdd = math.add(A, B)
console.timeEnd('add')

console.time('matrix map subset')
const matrixMapSub = addMatrixMapSubset(A, B)
console.timeEnd('matrix map subset')

console.time('matrix map get')
const matrixMap = addMatrixMapGet(A, B)
console.timeEnd('matrix map get')

console.time('flat map')
const matrixFlatMap = addFlatMap(A, B)

console.timeEnd('flat map')
console.log('matrix map subset OK', math.flatten(math.equal(arrayAdd, matrixMapSub)).every(x => x))
console.log('matrix map get OK', math.flatten(math.equal(arrayAdd, matrixMap)).every(x => x))
console.log('flat map OK', math.flatten(math.equal(arrayAdd, matrixFlatMap)).every(x => x))


function addMatrixMapSubset(A, B) {
  const Amat = math.matrix(A)
  const Bmat = math.matrix(B)
  return Amat.map((a, id) => math.addScalar(a, Bmat.subset(math.index(...id)))).toArray()
}

function addMatrixMapGet(A, B) {
  const Amat = math.matrix(A)
  const Bmat = math.matrix(B)
  return Amat.map((a, id) => math.addScalar(a, Bmat.get(id))).toArray()
}

function addFlatMap(A, B) {
  const Aflat = math.flatten(A)
  const Bflat = math.flatten(B)
  return math.reshape(Aflat.map((a, id) => math.addScalar(a, Bflat[id])), math.size(A))
}
```

**Comment by josdejong** [comment]
Thanks for the updates David!

I'm surpised to see that Firefox performs so bad compared to Chrome, I would have expected them to be more or less in the same order of magnitude. Anyway, I'm happy you found that using `get` instead of `subset` helps so much! I think the performance is good enough, at least for now. We can always improve further but let's keep this PR focussed on the new feature itself - as soon as I get into optimization-mode it's hard for me to stop 😉.

Switching to matrices that are flat under the hood is a long-time wish, see https://github.com/josdejong/mathjs/discussions/2702 😄.

**Comment by dvd101x** [comment]
Thanks Jos!

About Firefox, I might be doing something wrong. I was testing straight from the console and maybe some developing stuff was kicking in. I will try something else and let you know.

Thanks I will continue with `get`, mostly as is. I will focus on that and clean up a bit.

Thanks a lot for the reference to flat dense matrices! I will read the thread and see if I can contribute somehow as I was doing some tests that found interesting.

I will focus on this PR just to add the new functionality and maybe later check for performance by flattening matrices.

Here is a preview of something I'm testing.

https://observablehq.com/@dvd101x/multiple-map-tests

**Comment by josdejong** [comment]
I got similar results on Firefox so I don't think you're doing something wrong there.

**Comment by dvd101x** [comment]
Thanks,

That's odd. By testing an html file I got something very different.

| Method | Chrome [ms] | Edge [ms]  | Firefox [ms] |
| -------|--------------------|-------------------|---------------------|
| add    | 55 | 53 | 57 |
| matrix map with get | 233 | 231 | 147 |
| flat map | 107 | 113 | 55 |

```html
< !DOCTYPE html >
  <html>
    <head>
      <title>This is a title</title>
    </head>
    <script src="https://unpkg.com/mathjs@13.0.3/lib/browser/math.js"></script>
    <body>
      <div>
        <p>Hello world!</p>
      </div>
    </body>
    <script defer>
      math.config({
        randomSeed: 100
})
      const A = math.random([1000, 1000])
      const B = math.random([1000, 1000])

      console.time('add')
      const arrayAdd = math.add(A, B)
      console.timeEnd('add')

      console.time('matrix map subset')
      const matrixMapSub = addMatrixMapSubset(A, B)
      console.timeEnd('matrix map subset')

      console.time('matrix map get')
      const matrixMap = addMatrixMapGet(A, B)
      console.timeEnd('matrix map get')

      console.time('flat map')
      const matrixFlatMap = addFlatMap(A, B)

      console.timeEnd('flat map')
console.log('matrix map subset OK', math.flatten(math.equal(arrayAdd, matrixMapSub)).every(x => x))
console.log('matrix map get OK', math.flatten(math.equal(arrayAdd, matrixMap)).every(x => x))
console.log('flat map OK', math.flatten(math.equal(arrayAdd, matrixFlatMap)).every(x => x))


      function addMatrixMapSubset(A, B) {
  const Amat = math.matrix(A)
      const Bmat = math.matrix(B)
  return Amat.map((a, id) => math.addScalar(a, Bmat.subset(math.index(...id)))).toArray()
}

      function addMatrixMapGet(A, B) {
  const Amat = math.matrix(A)
      const Bmat = math.matrix(B)
  return Amat.map((a, id) => math.addScalar(a, Bmat.get(id))).toArray()
}

      function addFlatMap(A, B) {
  const Aflat = math.flatten(A)
      const Bflat = math.flatten(B)
  return math.reshape(Aflat.map((a, id) => math.addScalar(a, Bflat[id])), math.size(A))
}
    </script>
  </html>
```

**Comment by josdejong** [comment]
😂  ha ha yeah I get similar results like you with this HTML file.

We should be careful with these benchmarks using `console.time`. It measures only once, which is a "cold start". It would be better to use a benchmark library that tests repeatedly and calculates the average and standard deviation, like in https://github.com/josdejong/mathjs/tree/develop/test/benchmark. 

Anyway, based on this rough benchmark I have good enough confidence that the performance of the multi map is "slower but in the same order of magnitude", so, perfectly fine for this first implementation 👌.

**Comment by dvd101x** [comment]
I found the issue that was affecting speed. Please check this on the new commit.

```js
const a = math.random([1000])
const b = math.random([1000, 1])

console.time('add')
math.add(a, b)
console.timeEnd('add')            // 1003 ms

console.time('map typed')
math.map(a, b, math.addScalar)
console.timeEnd('map typed')      // 862 ms

console.time('map function')
math.map(a, b, (x, y) => math.addScalar(x, y))
console.timeEnd('map function')   // 897 ms
```

**Comment by josdejong** [comment]
Ow nice 😎 

I get something like this:

```
// Chrome:
add: 741.507080078125 ms
map typed: 1483.096923828125 ms
map function: 638.359130859375 ms

// Firefox
add: 1996ms - timer ended
map typed: 2915ms - timer ended
map function: 2083ms - timer ended

// Node.js v20
add: 954.098ms
map typed: 2.316s
map function: 1.430s
```

Anyway: the performance of using multi map is close to not using it, which is really nice and better than I had expected!

**Comment by dvd101x** [comment]
Thanks Jos!

I included some tests and added some documentation in matrices.

I will try the next few days to include the type definitions.

**Comment by dvd101x** [comment]
Hi, it's ready for review.

**Comment by josdejong** [comment]
That sounds good, thanks for the updates! I'll look into it after the holidays (sorry for the delay).

**Comment by dvd101x** [comment]
Thank you Jos!
happy holidays!

File: .DS_Store
Comment (josdejong) [inline .DS_Store:1]: I think we should not add `.DS_Store` to the repo. Instead, I think we can list the file in `.gitignore` to prevent it from being committed?

File: test/.DS_Store
Comment (josdejong) [inline test/.DS_Store:1]: Here another `.DS_Store` file, can you remove that from the PR?

File: test/unit-tests/function/matrix/map.test.js
Comment (josdejong) [inline test/unit-tests/function/matrix/map.test.js:19]: Thanks for the extensive testing!

I think we also need some test for the map.transform.js version of the function, or am I overlooking them?

File: src/function/matrix/map.js
Comment (josdejong) [inline src/function/matrix/map.js:26]: How about documenting the syntax like:

```
math.map(x, callback)
math.map(x, y, ..., callback)
```

? 

Then it is clear (I think) that you can also pass more than two arrays/matrices.

File: src/function/matrix/map.js
Comment (josdejong) [inline src/function/matrix/map.js:28]: Can you also update the syntax and examples in the embedded docs file [/src/expression/embeddedDocs/function/matrix/map.js](https://github.com/josdejong/mathjs/blob/develop/src/expression/embeddedDocs/function/matrix/map.js)?

**Review by josdejong — CHANGES_REQUESTED** [review]
I've reviewed the latest version @dvd101x . It's really impressive how it looks now compared to your first draft: the code is now easy to follow, good performance, all is well documented and tested. Nice job!!

I added a few minor inline comments, can you have a look at those? After that we can merge this new feature :)

File: .DS_Store
Comment (dvd101x) [reply .DS_Store:1]: Hi Jos, yes of course. I didn't know what they were.

File: test/unit-tests/function/matrix/map.test.js
Comment (dvd101x) [reply test/unit-tests/function/matrix/map.test.js:19]: They are included in map.test.js the ones that state they should work from the parser from line 249 to 288. Is that ok?

**Comment by dvd101x** [comment]
Hi Jos, thank you for the thorough review and your kind words.

It wouldn't be possible without your help and patience.

The comments are resolved now, except for one.

While looking into optimization of the multiple map feature I saw an opportunity to improve speed with applyCallback that maybe could improve `forEach`, `filter` and others. I would like to address this in a separate discussion if I find a way to make it work.

I see there is also similar objective in #3251 but I think it could still be applied after that.

**Review by josdejong — APPROVED** [review]
All looks good, merging now.

**Comment by josdejong** [comment]
Published now in `v13.1.0`, thanks again David 🎉

**Comment by dvd101x** [comment]
Thank you Jos!

**Pull request merged**
